### ISSUE TITLE
Request_Denied when clicking on location

### ISSUE THREAD

**Issue by maxpac** [iam4x/pokemongo-webspoof#461]
Hi,

Facing this issue when I'm not able to click on a location to spoof (see photo)

![image](https://user-images.githubusercontent.com/45595233/49440959-1624d080-f800-11e8-9b0d-e5f070d0dfba.png)

Anyone facing the same issue?

**Comment by fritzggerard** [comment]
this might occur when the MAP API or Location API reached the cota limit. You have to add your own MAPs API key

**Comment by maxpac** [comment]
how do i do that?

**Comment by akattamis** [comment]
Having the same issue.

**Comment by MattyTesar** [comment]
Can be solved by using your own Maps API key, although this requires to setup a Google Cloud Platform project. 
After setting up the project, you can obtain the key here https://console.cloud.google.com/google/maps-apis/ 
Next clone the webspoof repository by following steps from Readme section "Develop", open src/config/api.js and replace the api key there. 
After that, you can run (`npm run dev`) or build (`npm run build`) the app for use with your own api key. (This should prevent the issue further, unless you manage to reach the api free quotas)

**Comment by molonlabe88** [comment]
Matty, can you dumb it down. I downloaded the nodejs but then what?

**Comment by MattyTesar** [comment]
For obtaining the api key, please follow this  https://developers.google.com/maps/documentation/javascript/get-api-key . You will be asked to create or link a billing account though, which requires you enter your credit card information. But if you'll be using this only for the webspoof app, no excessive use, you won't be billed anything. (To be sure you can setup a budget of $0 to be notified in case of any charge)
(Make sure that you enable Maps JavaScript API and Directions API - check here https://console.cloud.google.com/google/maps-apis/overview)

For the next steps, open terminal.app on your mac, check you have NodeJS installed (eg. by running `npm -v`. If you get "command not found", install nodejs from here https://nodejs.org/en/ ), then go to the folder where you want to clone this repository (by typing `cd 'path_to_folder'`), run `git clone https://github.com/iam4x/pokemongo-webspoof.git`, `cd pokemongo-webspoof` and `npm install`
Then proceed to replace the api key in path_to_folder/pokemongo-webspoof/src/config/api.js and save the file.
Then enter `npm run build` in terminal, and after it finishes, you can find freshly built pokemongo-webspoof.app in pokemongo-webspoof-darwin-x64 folder.

Is that understandable, or are you having trouble with any specific step/s?

**Comment by CavemanGer** [comment]
Hi, unfortunately Request denied even after replacing with own key. Any suggestions?

**Comment by megusta63** [comment]
Also the same "Request_Denied" fail after replacing the API Keys.. Ah its a shame if got an EX Raid on Friday :(

**Comment by MattyTesar** [comment]
Did you build the app after saving the changes and run the new version? Also, make sure you have Maps JavaScript or Embed API key with unrestricted use (you can check here https://console.cloud.google.com/google/maps-apis/overview - open the desired api detail from 'Enabled APIs' and then look to 'Credentials' section)

**Comment by CavemanGer** [comment]
For some reason the api key was not activated. Resolved this, now it works. Thanks a lot Matty!

**Comment by megusta63** [comment]
i dont get this, i replaced the API Key in the config file bust still have the same issue. maybe i can try with one of your API Keys?

**Comment by Blistex1899** [comment]
Thanks a lot Matty for your Help.
After replacing the API Key and build the app the error Google Maps API error 400 has gone.
But "Request Denied" is still their. Which API ´s has to be activated?
Thank´s

**Comment by Blistex1899** [comment]
Direction API has also be activated, after that it works fine.

Thanks

**Comment by MattyTesar** [comment]
> Direction API has also be activated, after that it works fine.

You're right, Directions API is quite essential for the autopilot feature. Thanks for pointing that out 👍

**Comment by megusta63** [comment]
I get error messages after these steps:
cd pokemongo-webspoof && npm install
npm run build

macbook-air:~ User$ CD pokemongo-webspoof
macbook-air:~ User$ CD pokemongo-webspoof && npm install
npm WARN saveError ENOENT: no such file or directory, open '/Users/xyz/package.json'
npm WARN enoent ENOENT: no such file or directory, open '/Users/xyz/package.json'
npm WARN xyz No description
npm WARN xyz No repository field.
npm WARN xyz No README data
npm WARN xyz No license field.

if i understand the errormessage right its because the package.json file cant be found, but when i look in the pokemongo-webspoof folder its inside. so i copied the file in the root directory (user).. after this the installation "worked" but still with warnings..

After this i searched for the pokemongo-webspoof-darwin-x64, but i cant find it.
i can find a node_modules but no pokemongo.webspoof in it.. 

isnt it possible to change the API Key in an existing Pokemon Webspoof APP? There should be an api.js file too..

sorry for this nooby questions but im new in MAC..

**Comment by MattyTesar** [comment]
@megusta63 these `npm install` warnings aren't really important, if you see node_modules in your pokemongo-webspoof folder then the installation probably worked. Did you run `npm run build` after that, and if so, do you get any errors there? (make sure you run this from inside the pokemongo-webspoof folder)

**Comment by maxpac** [comment]
i'm getting npm command not found. Sorry, i'm bad at codings and commands

**Comment by MattyTesar** [comment]
@maxpac have you installed NodeJS? You can try using absolute path to the command, which should be `/usr/local/bin/npm`
If you're still getting "command not found" or "No such file or directory", try installing node with the package from https://nodejs.org/en/

**Comment by maxpac** [comment]
sorry, forgot i had to install node.js. so i have followed all the steps, ensured my api has no restrictions but i'm still getting Request Denied,

I've also ensured both Map Javascript API and Directions API are enabled.

**Comment by maxpac** [comment]
Regarding this part "(Make sure that you enable Maps JavaScript API and Directions API - check here ", i don't get it. Isn't there just 1 API key?

**Comment by maxpac** [comment]
z

**Comment by megusta63** [comment]
@MattyTesar worked! thanks, i run from the root directory first because of this it doesnt worked

**Comment by maxpac** [comment]
@megusta63 Thanks for bringing this up. No wonder i couldn't get through.

However, after "cd pokemongo-webspoof and npm install", i'm getting ACTION Regenerating Makefile. What's this?

**Comment by MattyTesar** [comment]
@maxpac can you try removing node_modules folder and package.lock file and running `npm install` again (from inside pokemongo-webspoof folder). Also make sure you have nodejs version 6+ ( run `node -v` )

**Comment by maxpac** [comment]
@MattyTesar ok i'm able to to pass npm install. The problem right now is npm run build. I'm getting this error Failed at the pokemongo-webspoof@1.2.1 build script. Any idea?

The first error code line was ELIFECYCLE

**Comment by maxpac** [comment]
IT FINALLY WORKED! Had to run npm run build first, change the API key in api.js, run npm run build again. Not sure why an additional steps but it works at least.

Thanks again everyone! Especially @MattyTesar

**Comment by maxpac** [comment]
Sorry to trouble again. I'm able to build, run without issues, able to teleport but my character doesn't respond to my movements at all. I'm getting this error.

![image](https://user-images.githubusercontent.com/45595233/49524692-34b1c700-f8e7-11e8-841c-5afa1b19d0f4.png)

**Comment by MattyTesar** [comment]
Check Readme section Known issues 
https://github.com/iam4x/pokemongo-webspoof/blob/c820192db572c3829add43d34e00adc156c7e6de/README.md#known-issues

**Comment by maxpac** [comment]
Cause one of the troubleshooting was to delete the app and reinstall it. So does that mean i have do npm run build after deleting again?

**Comment by MattyTesar** [comment]
That shouldn't be necessary, unless you build the app with a different version number. You just need to add your build of pokemongo-webspoof.app to the accessibility apps in System Preferences > Security & Privacy > Privacy tab > Accessibility and check the checkbox next to it.
Or bypass this altogether by opening Script Editor.app and running the following code (if you do this, don't use "Auto update Xcode location" in the spoof app)
```
repeat
	tell application "System Events" to tell process "Xcode"
		click menu item "pokemonLocation" of menu 1 of menu item "Simulate Location" of menu 1 of menu bar item "Debug" of menu bar 1
	end tell
	delay 0.5
end repeat
```

**Comment by rudychidiac** [comment]
Hi there,
I'm sorry to ask but is there any free of cost solution to get Webspoof to work again like before?
Thank you

**Comment by MattyTesar** [comment]
You mean without the need to setup billing account? Because even with it, it's basically free of cost, unless you're planning on making more than about 1300 directions requests (autopilot target updates) per day.
https://cloud.google.com/maps-platform/pricing/sheet/

**Comment by molonlabe88** [comment]
I enabled Maps API and now maps load without the watermark but I am still getting the "request denied"

Well, it was loading fine. Now its back to the "development purposes only"

now i'm getting over_query_limit

**Comment by molonlabe88** [comment]
> You mean without the need to setup billing account? Because even with it, it's basically free of cost, unless you're planning on making more than about 1300 directions requests (autopilot target updates) per day.
> https://cloud.google.com/maps-platform/pricing/sheet/

or what about a way that would let us update the API without as many steps. I think i've almost got this time, but I don't follow what to do if i have to reinstall web spoof. I'd even donate a few bucks to someone who made it easier.

**Comment by k-webdesign** [comment]
Any chance to switch to openstreetmap or something else? @iam4x

**Comment by maxpac** [comment]
@MattyTesar  seems that rebuilding did the trick for me.

Anyway, what counts as 1 direction request? Say for instance if i walk from 1 place to another, maybe 10km, how many direction requests will that be?

**Comment by CavemanGer** [comment]
> I enabled Maps API and now maps load without the watermark but I am still getting the "request denied"
> 
> Well, it was loading fine. Now its back to the "development purposes only"
> 
> now i'm getting over_query_limit

Well, it was working fine for just one day. Now I've got also "over_query_limit" and watermark back again. Got a new key, rebuild, same thing. I'm a bit lost now...

**Comment by taylorsm** [comment]
> > I enabled Maps API and now maps load without the watermark but I am still getting the "request denied"
> > Well, it was loading fine. Now its back to the "development purposes only"
> > now i'm getting over_query_limit
> 
> Well, it was working fine for just one day. Now I've got also "over_query_limit" and watermark back again. Got a new key, rebuild, same thing. I'm a bit lost now...

Alright, so what APPEARS to have worked is that my billing account on google was not associated with a project/API. If you go into your API and look for quota, it will show you where you are hitting over the limit and if you just kinda follow how to increase the limit, then you will see. 

Doesn't appear as if running webspoof does very many requests.

**Comment by rudychidiac** [comment]
nothing is happening when I do the last step in terminal "npm run build"
I'm sorry i'm a noob when it comes to these stuff and I'm really trying to understand what i'm doing incorrectly.

Edit: nevermind, I just had to be patient and wait for it to finish. Everything works fine so far! Thank you so much @MattyTesar :)

**Comment by Dooch88** [comment]
And how can you fix this if you don't have . credit card? Are the MAp API's connected to your own computer? So a fresh re-install could also do the trick?

**Comment by Snoverfield** [comment]
Been having this issue for awhile and been just running routes with gpx files to work around this. But would be nice to be able to move across the map easier without tapping arrow keys or creating an entire gpx file each time.

Anywho, I'm a total noob at this and people on here have been super helpful. @MattyTesar I'm trying your fix and I got to the point of 

_> For the next steps, open terminal.app on your mac, check you have NodeJS installed (eg. by running `npm -v`. If you get "command not found", install nodejs from here https://nodejs.org/en/ ),_ **then go to the folder where you want to clone this repository (by typing `cd 'path_to_folder'`), run `git clone https://github.com/iam4x/pokemongo-webspoof.git`, `cd pokemongo-webspoof` and `npm install`**
_> Then proceed to replace the api key in path_to_folder/pokemongo-webspoof/src/config/api.js and save the file.
> Then enter `npm run build` in terminal, and after it finishes, you can find freshly built pokemongo-webspoof.app in pokemongo-webspoof-darwin-x64 folder.
> 
> Is that understandable, or are you having trouble with any specific step/s?_

Do I type exactly 'path_to_folder' or do I need to manually enter an actual folder name. I  copied what you had exactly and no such folder or directory was found. And how do I find the folder I wanna use. This step I'm totally lost. Any help or explanation would be appreciated. Thanks!
